src/node/upload.ts

```diff
@@ -0,0 +1,100 @@
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { createReadStream, promises, statSync } from 'fs';
+import { join } from 'path';
+
+import {
+  TurboAuthenticatedBaseUploadService,
+  defaultUploadServiceURL,
+} from '../common/upload.js';
+import {
+  TurboAuthenticatedUploadServiceConfiguration,
+  TurboUploadDataItemResponse,
+  TurboUploadFolderParams,
+  TurboUploadFolderResponse,
+  isNodeUploadFolderParams,
+} from '../types.js';
+
+export class TurboAuthenticatedNodeUploadService extends TurboAuthenticatedBaseUploadService {
+  constructor({
+    url = defaultUploadServiceURL,
+    retryConfig,
+    signer,
+    logger,
+    token,
+  }: TurboAuthenticatedUploadServiceConfiguration) {
+    super({ url, retryConfig, logger, token, signer });
+  }
+
+  private async getAbsoluteFilePathsFromFolder(
+    folderPath: string,
+  ): Promise<string[]> {
+    const absoluteFilePaths: string[] = [];
+
+    // Walk the directory and add all file paths to the array
+    const files = await promises.readdir(folderPath);
+    for (const file of files) {
+      const absoluteFilePath = join(folderPath, file);
+      const stat = await promises.stat(absoluteFilePath);
+      if (stat.isDirectory()) {
+        absoluteFilePaths.push(
+          ...(await this.getAbsoluteFilePathsFromFolder(absoluteFilePath)),
+        );
+      } else {
+        absoluteFilePaths.push(absoluteFilePath);
+      }
+    }
+    return absoluteFilePaths;
+  }
+
+  async uploadFolder(
+    params: TurboUploadFolderParams,
+  ): Promise<TurboUploadFolderResponse> {
+    if (!isNodeUploadFolderParams(params)) {
+      throw new Error('folderPath is required for node uploadFolder');
+    }
+    const { folderPath, indexFile, dataItemOpts, signal } = params;
+
+    const paths: Record<string, { id: string }> = {};
+    const fileResponses: TurboUploadDataItemResponse[] = [];
+
+    const absoluteFilePaths =
+      await this.getAbsoluteFilePathsFromFolder(folderPath);
+
+    for (const absoluteFilePath of absoluteFilePaths) {
+      const result = await this.uploadFile({
+        fileStreamFactory: () => createReadStream(absoluteFilePath),
+        fileSizeFactory: () => statSync(absoluteFilePath).size,
+        signal,
+        dataItemOpts,
+      });
+
+      fileResponses.push(result);
+      const relativePath = absoluteFilePath.replace(folderPath, '');
+      paths[relativePath] = { id: result.id };
+    }
```

Comment:
a few comments:
- consider doing this concurrently and in a p-limit
- consider how we want to handle one file falling after x number of retries. we may want to provide the option to abort on failure, e.g. `abortOnFailure` or `throwOnFailure` that throws the error of the last upload. alternatively, if `throwOnFailure` is `false` we catch any errors and return them with the result here. I'm guessing we do something similar in ardrive-js?

Reply:
110a424 got the pLimit part worked out here

Reply:
a258aa6 heres is a throw on failure commit, with errors being optionally returned on false